generator/camera_info_collector.hpp

```diff
@@ -0,0 +1,157 @@
+
+    bool IsValid() const { return m_version == CamerasInfoCollector::kLatestVersion; }
+
+  private:
+    uint32_t m_version = 0;
+    uint32_t m_amount = 0;
+  };
+
+  CamerasInfoCollector(std::string const & dataFilePath, std::string const & camerasInfoPath,
+                       std::string const & osmIdsToFeatureIdsPath);
+
+  static uint32_t constexpr kLatestVersion = 1;
+  static uint32_t constexpr kMaxCameraSpeedKmpH = std::numeric_limits<uint8_t>::max();
+
+  void Serialize(FileWriter & writer) const;
+
+  bool IsValid() const { return m_valid; }
+
+private:
+
+  struct Camera
+  {
+    // feature_id and segment_id, where placed camera.
+    // m_coef - coef from 0 to 1 where it placed at segment from the begin.
+    struct SegmentCoord
+    {
+      SegmentCoord() = default;
+      SegmentCoord(uint32_t fId, uint32_t sId, double k) : m_featureId(fId), m_segmentId(sId), m_coef(k) {}
+
+      uint32_t m_featureId = 0;
+      uint32_t m_segmentId = 0;
+      double m_coef = 0.0;
+    };
+
+    static double constexpr kEqualityEps = 1e-5;
+
+    Camera(m2::PointD const & center, uint8_t const maxSpeed, std::vector<SegmentCoord> && ways)
+      : m_center(center), m_maxSpeedKmPH(maxSpeed), m_ways(std::move(ways))
+    {
+    }
+
+    void ParseDirection()
+    {
+      // After some research we understood, that camera's direction is a random option.
+      // This fact was checked with random-choosed cameras and google maps 3d view.
+      // The real direction and the indicated in the .osm differed.
+      // So let's this function, space in mwm will be. And implementation won't, until
+      // well data appeared in .osm
+    }
+
+    // Modify m_ways vector. Set for each way - id of the closest segment.
+    void FindClosestSegment(FrozenDataSource const & dataSource, MwmSet::MwmId const & mwmId);
+
+    // Returns empty object, if current feature - |wayId| is not the road, and we should erase it
+    // from vector of |m_ways|.
+    // Otherwise returns id of segment from feature with id - |wayId|, which starts at camera's
+    // center and coefficient - where it placed at the segment.
+    boost::optional<std::pair<double, uint32_t>> FindMyself(
+      uint32_t wayId, FrozenDataSource const & dataSource, MwmSet::MwmId const & mwmId) const;
+
+    void Serialize(FileWriter & writer, uint32_t prevFeatureId) const;
+
+    m2::PointD m_center;
+    uint8_t m_maxSpeedKmPH;
+    std::vector<SegmentCoord> m_ways;
+    CameraDirection m_direction = CameraDirection::Unknown;
+  };
+
+  static double constexpr kMaxDistFromCameraToClosestSegmentMeters = 20.0;
+  static double constexpr kSearchCameraRadiusMeters = 10.0;
+
+  bool ParseIntermediateInfo(std::string const & camerasInfoPath,
+                             std::map<osm::Id, uint32_t> const & osmIdToFeatureId);
+
+  bool m_valid = true;
+  std::vector<Camera> m_cameras;
+};
+}  // namespace generator
+
+namespace routing
+{
+// To start build camera info, next data must be ready:
```

Comment:
s/build/building/
s/next/the following/

Reply:
done